### DISCUSSION TITLE
[Kubernetes] Executor slot count continues to fill up despite Pool Size maintaining steady

### DISCUSSION THREAD

**Discussion by kencieszykowski** [apache/airflow#37732] [Q&A]
Hello everyone,

First off, I'm moreso on the Administrative side on Airflow than the developer side-- so I have limited context to our DAGs and how/why a bad actor could be causing the situation we're seeing below, but it has been continually happening for the last few months.

Here are some details of our Airflow infra/setup:
- EKS v1.25
- Airflow 2.7.1 (deployed via community Helm chart)
- EFS for storing packaged DAGs
- EFS for logs
- RDS for Database connections
- 10+ Scheduler Replicas
- Parallelism set to 2048
- Default Pool set to 2048
- Exclusively using the KubernetesPodOperator

There are a few things that we're seeing that are a little strange-- and I'm trying to determine if this is a poorly configured DAG, or if this is related to our Airflow infrastructure, but I can't seem to get the metrics that I'm really looking for.  I have seen that some metrics around task instances aren't available with the Kubernetes Operator-- but I think we should have enough to work with.

We only have about 20-25 DAGs in our environment, but several of them can fan out into tens or even hundreds of task instances.  Many downstream tasks require successes upstream, and we have retries enabled almost everywhere.

Nevertheless, here is the problem:

The Executor Slots open (which seems to be mapped directly to Parallelism continually moves downwards.  Even though tasks are completing on the Kubernetes side (and marked properly in Airflow's UI), it seems like _something_ is taking up slots and not letting go of them.

We have a wildly overprovisioned database with a very large amount of connections available, but routinely don't hit more than 250 connections.  

I've attached a screenshot showing the behavior below.  We don't have any specific pools specified, so we would expect that any running task would grab a slot from the default_pool and then return it upon completion (we are seeing this behavior).

Problem is that the Executor slots are constantly decreasing and don't seem to be increasing when task instances are completed.  I've attached some screenshots to show this behavior.

Additionally, we're a little confused as the Airflow docs say that:
> This defines the maximum number of task instances that can run concurrently per scheduler in Airflow, regardless of the worker count. Generally this value, multiplied by the number of schedulers in your cluster, is the maximum number of task instances with the running state in the metadata database.

If this is set per schedule replica, I would imagine to be getting this metric emitted from all of the Scheduler hosts inside of Airflow-- but we're only getting one value from one host; 2048.  

Aside from dynamic DAGs and Task Instances, I've heard people talk about XCom and that we're using it extensively.  As mentioned earlier, I don't have a whole lot of context on what the DAGs are doing specifically, but I'm willing to learn-- especially if we can get over this issue here.

As far as the screenshots go, the first one is from Datadog that displays the Pool open slots vs. running slots.  This also has the Executor slot metrics in the row above it.  Another shows the `Cluster Activity` panel of Airflow2, and a third shows the amount of TIs that have started in a timeframe (with what I suspect is the catalyst to our problems)

Another one shows a misconfigured DAG that isn't pushing a `dag_id` -- and what I suspect is wreaking havoc over things here; I'm just at a loss for where to begin on debugging this.  All metrics and logs don't seem to be showing anything unexpected.  There aren't very many zombie tasks being killed, and there isn't any instability even when we concurrently run 500+ tasks.

Any and all guidance or messages are very appreciated.  Happy to provide any other information that would/could help out as well!
![capture 2024-02-26 at 12 21 41 PM](https://github.com/apache/airflow/assets/13755642/05c558dd-5d2e-448c-9831-7c01ef8a703a)
![capture 2024-02-26 at 2 44 56 PM](https://github.com/apache/airflow/assets/13755642/fc2a14ce-2b23-454e-ae09-b57d3c88853a)
![capture 2024-02-26 at 1 55 00 PM](https://github.com/apache/airflow/assets/13755642/7e664b39-176f-4867-abd5-075cbd0a236f)

**Comment by potiuk** [comment]
I believe there were related issues that have been addressed already in later versions of Airflow. We are at 2.8.3 right now and the best suggestion I might have is to upgrade to latest version. Note that in general - even if you were afraid to upgrade to 2.8 (though there is no reason to be afraid) - you have 2.7.1 and 2.7.3 is the latest version in 2.7 feature line - so it contains ONLY bgfixes, nothing else - so you miss on all those bufgfixes - including some security bugfixes. 

My suggestion - upgrade to 2.8.3 and see if the problem re-occurs there.

**Reply by potiuk** [reply]
And just to explain why this is the best approach.

Once you do that and it will continue to happen - it's best if you report it as issue then (but upgrading save a lot of time of your - and anyones time who would like to help you in their free time, as it will cut of all the time in case the issue has already been fixed (which is likely - every release here is few tens of bugs fixed. 

There is no-one who would be able to track all the issues and be able to authoritatively say "it's fixed" - it does ring a bell but not enough to be able to pin-point concrete issue and concrete solution. And since poeple who contribute here are volunteers, they will hesitate to spend their time on looking through hundreds of issues to find out the exact one that causes it.  You can try to do it yourself - the github issue and discussions are pretty well searchable and maybe you will find an issue that will be marked as solved.

But if you do upgrade to latest version then it at least cut out on the possibility that it's aldready solved, It has much higher chance then that a maintainer - who will know it happens on the latest version will spend their (again - volunteer - time) on analysing and diagnosing it .

**Reply by kencieszykowski** [reply]
Right on!  Appreciate the detailed response here.  I know that some of our Devs have been testing 2.8 locally and were having some issues with the bundled libraries.  No reason not to push to the latest version of 2.7 though.  Will update soon, and add any more details here.

Thank you for talking in a detailed and focused way here-- and appreciate your work!

**Reply by potiuk** [reply]
Do - let us know about the findings!

**Reply by potiuk** [reply]
BTW. @kencieszykowski - you might actually want to try Airflow 2.8.2 released yesterday. One of the "bundling" problem you could have is that Airflow started to depend on Pydantic > 2 and in 2.8.2 we relaxed it - and pydantic is no longer a requirement. That could be the reason why you would not have it easy to upgrade also for Airflow 2.7 - I think Pydantic >= 2 was introduced some time in 2.7.* line, so - surprisingly - it might be easier to migrate to Airlfow 2.8.2 than to 2.7.3

**Reply by kencieszykowski** [reply]
Ah-- looks like it wasn't related to Pydantic; instead here is a response I got:
> A DAG that used a kwarg that matched a renderable variable, and they were adding a default arg and then it was complaining about kwarg being define before args, nothing to do with pydantic, just the decorator.  
Example:
`def example(foo, bar):`
was being converted (assuming foo is a renderable value, but bar isn't) to:
`def example(foo=None, bar):`
but the call stack was beyond cryptic on the error.

Looks like they've patched up the issues, but unfortunately our Airflow updates got deprioritized.  Will update the thread when we get around to it to see if behavior changes with regards to tasks stacking.  Thanks again!

Edit: Came across this Issue:
https://github.com/apache/airflow/issues/32928

Which backs up the 'upgrade' option to gather that new CNCF Kubernetes plugin to 7.12+.  Will report back when we get this upgraded-- hopefully with good news 🤞

**Reply by potiuk** [reply]
Ah yes. We had another one about this one reported today: https://apache-airflow.slack.com/archives/CCQ7EGB1P/p1709136900917279

The reason here: https://github.com/apache/airflow/pull/33430

cc: @uranusjr -> just thinking if there is at least a way to make it more obvious :(
